toolchain/check/check_unit.cpp

```diff
@@ -511,9 +514,52 @@ auto CheckUnit::CheckRequiredDefinitions() -> void {
   }
 }
 
+auto CheckUnit::CheckPoisonedConcreteImplLookupQueries() -> void {
+  // Impl lookup can generate instructions (via deduce) which we don't use, as
+  // we're only generating diagnostics here, so we catch and discard them.
+  context_.inst_block_stack().Push();
+  auto poisoned_queries =
+      std::exchange(context_.poisoned_concrete_impl_lookup_queries(), {});
+  for (const auto& poison : poisoned_queries) {
+    auto witness_result =
+        EvalLookupSingleImplWitness(context_, poison.loc_id, poison.query,
+                                    poison.non_canonical_query_self_inst_id,
+                                    /*poison_concrete_results=*/false);
+    CARBON_CHECK(witness_result.has_concrete_value());
+    auto witness_id = witness_result.concrete_witness();
+    if (witness_id != poison.impl_witness) {
+      // We can get the `Impl` from the resulting witness here, which is the
+      // `Impl` that conflicts with the previous poison query.
+      auto table_id = context_.insts()
+                          .GetAs<SemIR::ImplWitness>(witness_id)
+                          .witness_table_id;
+      auto impl_id =
+          context_.insts().GetAs<SemIR::ImplWitnessTable>(table_id).impl_id;
+      const auto& impl = context_.impls().Get(impl_id);
+      auto interface_name_id =
+          context_.interfaces().Get(impl.interface.interface_id).name_id;
+      CARBON_DIAGNOSTIC(
+          PoisonedImplLookupConcreteResult, Error,
+          "found `impl` that changes the result of a previous use "
+          "of interface `{0}`",
+          SemIR::NameId);
+      auto builder =
+          emitter_.Build(impl.first_decl_id(), PoisonedImplLookupConcreteResult,
+                         interface_name_id);
+      CARBON_DIAGNOSTIC(PoisonedImplLookupConcreteResultNote, Note,
+                        "use of interface `{0}` here", SemIR::NameId);
+      builder.Note(poison.loc_id, PoisonedImplLookupConcreteResultNote,
+                   interface_name_id);
```

Comment:
I think it'd be good to list the full self type and interface type from the query here. Perhaps something like:
```suggestion
      CARBON_DIAGNOSTIC(
          PoisonedImplLookupConcreteResult, Error,
          "found `impl` that changes the result of a previous use "
          "of `{0} as {1}`",
         InstIdAsType, InstIdAsType);
      auto builder =
          emitter_.Build(impl.first_decl_id(), PoisonedImplLookupConcreteResult,
                         query.query_self_inst_id, query.query_specific_interface_id);
      CARBON_DIAGNOSTIC(PoisonedImplLookupConcreteResultNote, Note,
                        "impl used here");
      builder.Note(poison.loc_id, PoisonedImplLookupConcreteResultNote);
```
(And I think it's better to not repeat the same information in the note.)

Reply:
This won't quite work as is - the specific interface is not an instruction. The facet type instruction that is the constraint is not around at this point - it was decomposed into individual interfaces, and I want to avoid remaking a FacetType instruction in here, but using the intended ideas here.